src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -230,31 +233,43 @@ makePacketSender ::
   (ByteString -> m ()) ->
   ByteString ->
   m ()
-makePacketSender limit options publish message 
+makePacketSender limit options publish message
   | limit <= 0 || ByteString.length message < limit =
     -- In the case that the maximum packet payload length @size@ is not at least
     -- 1 byte or the provided 'ByteString' @bytes@ is empty, then yield one
     -- terminal packet.
     let packet :: Packet ByteString
-        packet = (Packet message (PacketInfo 0 1))
-      in publish (wireWrapPacket' options packet)
-  | otherwise = 
-    -- Allocates a vector of containing @ByteString.length bytes / size@
-    -- elements rounded upwards. For example, fix the length of @bytes@ to
-    -- be 5 and @size@ to be 2, then a vector of length 3 is allocated.
-    let count :: Int
-        count = (ByteString.length message + (limit - 1)) `quot` limit
-     in forConcurrently_ [0 .. count - 1] \i -> do 
+        packet = Packet message (PacketInfo 0 1)
+     in publish (wireWrapPacket' options packet)
+  | otherwise = do
+    caps <- liftIO getNumCapabilities
+    jobs <- newTVarIO 0
+
+    replicateConcurrently_ caps $ fix \next -> 
```

Comment:
Just a thought. Since we are using QoS1 on all of our publishes, each of these threads will have to wait for an ack packet from the MQTT broker. So I would expect there to still be performance improvements from increased concurrency here beyond the number of capabilities. I don't know what that optimal number would be. It would probably take some benchmarking to dial that in.

Reply:
This is a digression off of this PR but Matt's comment made me think: can we make QoS tunable? I'm not sure if it's a good idea to use QoS0 on the appliance in grpc-gateway (maybe we can tolerate some message loss for the benefit of better performance?) but I was wondering if we might want to use QoS0 in grpc-rest-cloud?

Reply:
QoS0 is at most once delivery. Won't that cause us to lose some packets and possible data corruption (or worse silent truncation)? That also makes me wonder if QoS1 is safe, it can deliver duplicate messages.

Reply:
@ixmatus, we could make it tunable, but I don't see how we could tolerate dropped packets in a gRPC transaction?

Reply:
@rkaippully, QoS1 should be fine since we handle duplicate packets gracefully internally.

Reply:
> @rkaippully, QoS1 should be fine since we handle duplicate packets gracefully internally.

That's good, thanks for clearing that up!